File: dart/exercises-(in_kotlin)/exercise1_zoo.kt
// ============================================================
// SE 3242: Android Application Development
// Week 2 - Kotlin Exercises
// Exercise 1: Model a Zoo
// Student: Precious [Last Name]
// Date: March 2026
// Description: Demonstrates abstract classes, inheritance,
//              and polymorphism in Kotlin using a Zoo example.
// ============================================================

// ── Abstract base class ──────────────────────────────────────

/**
 * Abstract representation of an animal in the zoo.
 *
 * Concrete subclasses must supply the number of [legs] and implement
 * [makeSound] to return the animal's characteristic sound.
 *
 * @property name The animal's name (e.g. "Rex", "Tweety").
 */
abstract class Animal(val name: String) {

    /**
     * The number of legs this animal has.
     * Implemented by each concrete subclass.
     */
    abstract val legs: Int

    /**
     * Returns the sound this animal makes (e.g. `"Woof!"`, `"Meow!"`).
     */
    abstract fun makeSound(): String

    /**
     * Returns a human-readable description of the animal combining its
     * name, leg count and characteristic sound.
     *
     * Example: `"Rex has 4 legs and says Woof!"`
     *
     * @return Formatted description string.
     */
    fun describe(): String = "$name has $legs legs and says ${makeSound()}"

    /**
     * Returns a concise string representation of this animal, useful
     * for printing inside collections (e.g. `"Dog(Rex)"`).
     */
    override fun toString(): String = "${this::class.simpleName}($name)"
}

// ── Concrete subclasses ──────────────────────────────────────

/**
 * A domestic dog with four legs that barks.
 *
 * @param name The dog's name.
 */
class Dog(name: String) : Animal(name) {
    /** Dogs have 4 legs. */
    override val legs: Int = 4

    /** Returns `"Woof!"`. */
    override fun makeSound(): String = "Woof!"
}

/**
 * A domestic cat with four legs that meows.
 *
 * @param name The cat's name.
 */
class Cat(name: String) : Animal(name) {
    /** Cats have 4 legs. */
    override val legs: Int = 4

    /** Returns `"Meow!"`. */
    override fun makeSound(): String = "Meow!"
}

/**
 * A bird with two legs that chirps.
 *
 * @param name The bird's name.
 */
class Bird(name: String) : Animal(name) {
    /** Birds have 2 legs. */
    override val legs: Int = 2

    /** Returns `"Tweet!"`. */
    override fun makeSound(): String = "Tweet!"
}

// ── Entry point ──────────────────────────────────────────────

/**
 * Demonstrates abstract classes, inheritance and polymorphism by
 * creating a mixed list of [Animal] instances and iterating over
 * them in several different ways.
 */
fun main() {
    val separator = "─".repeat(50)

    // Create a mixed list of animals
    val zoo: List<Animal> = listOf(
        Dog("Rex"),
        Cat("Whiskers"),
        Bird("Tweety"),
        Dog("Buddy"),
        Cat("Luna"),
        Bird("Polly"),
        Dog("Max")
    )

    // ── Section 1: for loop — print each animal's sound ──────
    println("Section 1: Animal Sounds (for loop)")
    println(separator)
    for (animal in zoo) {
        println("${animal.name} says: ${animal.makeSound()}")
    }

    println()
    println(separator)
    println()

    // ── Section 2: forEach — print full describe() ───────────
    println("Section 2: Full Descriptions (forEach)")
    println(separator)
    zoo.forEach { animal ->
        println(animal.describe())
    }

    println()
    println(separator)
    println()

    // ── Section 3: filterIsInstance — dogs only ───────────────
    println("Section 3: Dogs Only (filterIsInstance<Dog>)")
    println(separator)
    val dogs: List<Dog> = zoo.filterIsInstance<Dog>()
    dogs.forEach { dog ->
        println(dog)
    }

    println()
    println(separator)
    println("Total animals : ${zoo.size}")
    println("Dogs          : ${dogs.size}")
    println("Cats          : ${zoo.filterIsInstance<Cat>().size}")
    println("Birds         : ${zoo.filterIsInstance<Bird>().size}")
    println(separator)
}

File: dart/exercises-(in_dart)/exercise1_zoo.dart
// ============================================================
// SE 3242: Android Application Development
// Week 2 - Dart Exercises
// Exercise 1: Model a Zoo
// Student: [Your Name]
// Date: March 2026
// Description: Demonstrates abstract classes, inheritance,
//              and polymorphism in Dart using a Zoo example.
// ============================================================

// ─────────────────────────────────────────────────────────────
// Abstract base class
// ─────────────────────────────────────────────────────────────

/// Abstract base class representing any animal in the zoo.
///
/// Concrete subclasses must implement [legs] and [makeSound].
/// The concrete method [describe] is shared by all animals.
abstract class Animal {
  /// The animal's name (e.g. "Rex", "Whiskers").
  final String name;

  /// Creates an [Animal] with the given [name].
  Animal(this.name);

  /// Returns the number of legs this animal has.
  int get legs;

  /// Returns the sound this animal makes as a [String].
  String makeSound();

  /// Returns a human-readable description using [name], [legs], and [makeSound].
  ///
  /// Example: `"Rex has 4 legs and says Woof!"`
  String describe() => '$name has $legs legs and says ${makeSound()}';

  /// Returns a concise debug representation of the animal.
  @override
  String toString() => '${runtimeType}(name: $name, legs: $legs)';
}

// ─────────────────────────────────────────────────────────────
// Concrete subclasses
// ─────────────────────────────────────────────────────────────

/// A [Dog] — a four-legged animal that barks.
class Dog extends Animal {
  /// Creates a [Dog] with the given [name].
  Dog(super.name);

  /// Dogs have 4 legs.
  @override
  int get legs => 4;

  /// Dogs say "Woof!".
  @override
  String makeSound() => 'Woof!';
}

/// A [Cat] — a four-legged animal that meows.
class Cat extends Animal {
  /// Creates a [Cat] with the given [name].
  Cat(super.name);

  /// Cats have 4 legs.
  @override
  int get legs => 4;

  /// Cats say "Meow!".
  @override
  String makeSound() => 'Meow!';
}

/// A [Bird] — a two-legged animal that chirps.
class Bird extends Animal {
  /// Creates a [Bird] with the given [name].
  Bird(super.name);

  /// Birds have 2 legs.
  @override
  int get legs => 2;

  /// Birds say "Tweet!".
  @override
  String makeSound() => 'Tweet!';
}

// ─────────────────────────────────────────────────────────────
// Entry point
// ─────────────────────────────────────────────────────────────

/// Demonstrates abstract classes, inheritance, and polymorphism
/// by building a small zoo and performing various list operations.
void main() {
  // ── Create a mixed list of animals ────────────────────────
  final List<Animal> zoo = [
    Dog('Rex'),
    Cat('Whiskers'),
    Bird('Tweety'),
    Dog('Buddy'),
    Cat('Luna'),
    Bird('Sky'),
  ];

  final separator = '─' * 50;

  // ── Section 1: for loop — print each animal's sound ───────
  print(separator);
  print('Section 1 — Each animal\'s sound (for loop)');
  print(separator);
  for (final animal in zoo) {
    print('${animal.name}: ${animal.makeSound()}');
  }

  print('');

  // ── Section 2: forEach — print full describe() ────────────
  print(separator);
  print('Section 2 — Full description (forEach)');
  print(separator);
  zoo.forEach((animal) => print(animal.describe()));

  print('');

  // ── Section 3: whereType<Dog>() — filter only dogs ────────
  print(separator);
  print('Section 3 — Dogs only (whereType<Dog>)');
  print(separator);
  final dogs = zoo.whereType<Dog>().toList();
  dogs.forEach((dog) => print(dog.toString()));

  print(separator);
}
